jms/org.wso2.carbon.transport.jms/src/main/java/org/wso2/carbon/transport/jms/factory/CachedJMSConnectionFactory.java

```diff
@@ -0,0 +1,277 @@
+                cachedConnection.close();
+            }
+            return true;
+        } catch (JMSException e) {
+            logger.error("JMS Exception while closing the connection.");
+        }
+        return false;
+    }
+
+    @Override public void closeConnection(Connection connection) {
+        closeConnection(connection, false);
+    }
+
+    @Override public void closeMessageConsumer(MessageConsumer messageConsumer) {
+        closeConsumer(messageConsumer, false);
+    }
+
+    @Override public void closeMessageProducer(MessageProducer messageProducer) {
+        closeProducer(messageProducer, false);
+    }
+
+    @Override public void closeSession(Session session) {
+        closeSession(session, false);
+    }
+
+    private boolean closeConnection(Connection connection, boolean forcefully) {
+        try {
+            if (this.cacheLevel < JMSConstants.CACHE_CONNECTION || forcefully) {
+                connection.close();
+            }
+        } catch (JMSException e) {
+            logger.error("JMS Exception while closing the connection.");
+        }
+        return false;
+    }
+
+    private boolean closeConsumer(MessageConsumer messageConsumer, boolean forcefully) {
+        try {
+            if (this.cacheLevel < JMSConstants.CACHE_CONSUMER || forcefully) {
+                messageConsumer.close();
+            }
+        } catch (JMSException e) {
+            logger.error("JMS Exception while closing the consumer.");
+        }
+        return false;
+    }
+
+    private boolean closeProducer(MessageProducer messageProducer, boolean forcefully) {
+        try {
+            if (this.cacheLevel < JMSConstants.CACHE_PRODUCER || forcefully) {
+                messageProducer.close();
+            }
+        } catch (JMSException e) {
+            logger.error("JMS Exception while closing the producer.");
+        }
+        return false;
+    }
+
+    private boolean closeSession(Session session, boolean forcefully) {
+        try {
+            if (this.cacheLevel < JMSConstants.CACHE_SESSION || forcefully) {
+                session.close();
+            }
+        } catch (JMSException e) {
+            logger.error("JMS Exception while closing the consumer.");
+        }
+        return false;
+    }
+
+    private void resetCache() {
+        if (cachedMessageConsumer != null) {
+            try {
+                cachedMessageConsumer.close();
+            } catch (JMSException e) {
+            }
+            cachedMessageConsumer = null;
+        }
+        if (cachedMessageProducer != null) {
+            try {
+                cachedMessageProducer.close();
+            } catch (JMSException e) {
+            }
+            cachedMessageProducer = null;
+        }
+        if (cachedSession != null) {
+            try {
+                cachedSession.close();
+            } catch (JMSException e) {
+            }
+            cachedSession = null;
+        }
+        if (cachedConnection != null) {
```

Comment:
This is a same as the previous block. Will lead to an error situation. Please correct